File: src/test/java/io/protobase/f7/formulas/math/ABSTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class ABSTest extends TestFormula {
  @Test
  public void testApply() {
    assertThat(ABS.SELF.apply(null, 10.0)).isEqualTo(10.0);
    assertThat(ABS.SELF.apply(null, -10.0)).isEqualTo(10.0);
    assertThat(ABS.SELF.apply(null, 0.0)).isEqualTo(0.0);
    assertThat(ABS.SELF.apply(null, -218637221.22)).isEqualTo(218637221.22);
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(ABS.SELF.apply(null, "10.0")).isEqualTo(10.0);
  }

  @Test
  public void testApply_grid() {
    assertThat(ABS.SELF.apply(null,
        Grid.builder().add(0, 0, -4.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(4.0);
  }

  @Test
  public void test_lookup() {
    ABS F = new ABS(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(-2.0);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(2.0);
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(ABS.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(ABS.SELF.apply(null, "A", "Too many")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/math/ABSTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { ABS } from "../../../../main/js/formulas/math/ABS";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("ABS", function () {
  it("should work with numbers", function () {
    assert.equal(ABS.SELF.run(null, 10), 10);
    assert.equal(ABS.SELF.run(null, -10), 10);
    assert.equal(ABS.SELF.run(null, -0), 0);
  });

  it("should work with strings", function () {
    assert.equal(ABS.SELF.run(null, "10"), 10);
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(ABS.SELF.run(null, new ValueException()), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new ABS(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(-2);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), 2);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, -4).add(0, 1, "A").build();
    assert.deepEqual(ABS.SELF.run(null, one), 4);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((ABS.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual((ABS.SELF.run(null, "A", "B") as NAException).name, F7ExceptionName.NA);
  });
});
